libp2p/Host.cpp

```diff
@@ -60,930 +60,938 @@ ReputationManager::ReputationManager()
-			if (j->isConnected())
-				ret.push_back(j->info());
-	return ret;
+    std::vector<PeerSessionInfo> ret;
+    RecursiveGuard l(x_sessions);
+    for (auto& i: m_sessions)
+        if (auto j = i.second.lock())
+            if (j->isConnected())
+                ret.push_back(j->info());
+    return ret;
 }
 
 size_t Host::peerCount() const
 {
-	unsigned retCount = 0;
-	RecursiveGuard l(x_sessions);
-	for (auto& i: m_sessions)
-		if (std::shared_ptr<SessionFace> j = i.second.lock())
-			if (j->isConnected())
-				retCount++;
-	return retCount;
+    unsigned retCount = 0;
+    RecursiveGuard l(x_sessions);
+    for (auto& i: m_sessions)
+        if (std::shared_ptr<SessionFace> j = i.second.lock())
+            if (j->isConnected())
+                retCount++;
+    return retCount;
 }
 
 void Host::run(boost::system::error_code const&)
 {
-	if (!m_run)
-	{
-		// reset NodeTable
-		DEV_GUARDED(x_nodeTable)
-			m_nodeTable.reset();
-
-		// stopping io service allows running manual network operations for shutdown
-		// and also stops blocking worker thread, allowing worker thread to exit
-		m_ioService.stop();
-
-		// resetting timer signals network that nothing else can be scheduled to run
-		DEV_GUARDED(x_runTimer)
-			m_timer.reset();
-
-		m_timerReset.notify_all();
-		return;
-	}
-
-	if (auto nodeTable = this->nodeTable()) // This again requires x_nodeTable, which is why an additional variable nodeTable is used.
-		nodeTable->processEvents();
-
-	// cleanup zombies
-	DEV_GUARDED(x_connecting)
-		m_connecting.remove_if([](std::weak_ptr<RLPXHandshake> h){ return h.expired(); });
-	DEV_GUARDED(x_timers)
-		m_timers.remove_if([](std::shared_ptr<boost::asio::deadline_timer> t)
-		{
-			return t->expires_from_now().total_milliseconds() < 0;
-		});
-
-	keepAlivePeers();
-	
-	// At this time peers will be disconnected based on natural TCP timeout.
-	// disconnectLatePeers needs to be updated for the assumption that Session
-	// is always live and to ensure reputation and fallback timers are properly
-	// updated. // disconnectLatePeers();
-
-	// todo: update peerSlotsAvailable()
-	
-	list<shared_ptr<Peer>> toConnect;
-	unsigned reqConn = 0;
-	{
-		RecursiveGuard l(x_sessions);
-		for (auto const& p: m_peers)
-		{
-			bool haveSession = havePeerSession(p.second->id);
-			bool required = p.second->peerType == PeerType::Required;
-			if (haveSession && required)
-				reqConn++;
-			else if (!haveSession && p.second->shouldReconnect() && (!m_netPrefs.pin || required))
-				toConnect.push_back(p.second);
-		}
-	}
-
-	for (auto p: toConnect)
-		if (p->peerType == PeerType::Required && reqConn++ < m_idealPeerCount)
-			connect(p);
-	
-	if (!m_netPrefs.pin)
-	{
-		unsigned pendingCount = 0;
-		DEV_GUARDED(x_pendingNodeConns)
-			pendingCount = m_pendingPeerConns.size();
-		int openSlots = m_idealPeerCount - peerCount() - pendingCount + reqConn;
-		if (openSlots > 0)
-			for (auto p: toConnect)
-				if (p->peerType == PeerType::Optional && openSlots--)
```

Comment:
Here I guess the intention was to break the loop when we run out of `openSlots`. What was really happening is of course `openSlots` going negative when there are a lot of peers in `toConnect` available and it tried to connect to all of them at once.

Then `connect()` added all these peers to `m_pendingPeerConns` (the set of peers we are trying to connect to) and this variable affects `peerSlotsAvailable()`, so when the handshake was complete with any peer it checked `peerSlotsAvailablr()` (in `startPeerSession()`) and it always looked as no available slots and so it disconnected.